services/src/main/java/org/keycloak/protocol/ProtocolMapperUtils.java

```diff
@@ -83,25 +86,52 @@ public class ProtocolMapperUtils {
     // Script mapper goes last, so it can access the roles in the token
     public static final int PRIORITY_SCRIPT_MAPPER = 50;
 
-    public static String getUserModelValue(UserModel user, String propertyName) {
+    public static final String USER_PROPERTY_MAPPER_WRONG_PROPERTY = "userPropertyMapperWrongProperty";
+
+    private static final Map<String, Method> USER_MODEL_METHODS = new HashMap<>();
+    static {
+        for (Method declaredMethod : UserModel.class.getDeclaredMethods()) {
+            if (declaredMethod.getParameterCount() == 0 && !(declaredMethod.getReturnType().isAssignableFrom(Stream.class)) && !(declaredMethod.getReturnType().isAssignableFrom(Map.class))) {
+                if (declaredMethod.getName().startsWith("is")) {
+                    memorizedMethod(declaredMethod.getName().substring(2), declaredMethod);
+                } else if (declaredMethod.getName().startsWith("get")) {
+                    memorizedMethod(declaredMethod.getName().substring(3), declaredMethod);
+                }
+            }
+        }
+    }
 
-        String methodName = "get" + Character.toUpperCase(propertyName.charAt(0)) + propertyName.substring(1);
-        try {
-            Method method = UserModel.class.getMethod(methodName);
-            Object val = method.invoke(user);
-            if (val != null) return val.toString();
-        } catch (Exception ignore) {
+    private static void memorizedMethod(String property, Method method) {
+        property = property.substring(0, 1).toLowerCase() + property.substring(1);
+        Method existingMethod = USER_MODEL_METHODS.put(property, method);
```

Comment:
After this commit, we break existing mappers that are configured with the first letter upper-cased.

What if we do the migration in two steps? 
1. step
- We will change this PR to work also with the uppercase first letter (in this method we associate the method with both `property` and `property.substring(0, 1).toLowerCase() + property.substring(1)`). 
- The validation will be changed to work only with the lowercased first letter so new configurations are correct
- `getUserModelValue` prints a warning if the uppercased version is used
2. step - uppercased version does not work at all
WDYT?

Reply:
I'd like to keep a clean and efficient code in Keycloak. By supporting both, it will make the code harder to understand. I'd therefore suggest to move it to a migration so that at least data in a database is migrated automatically on a Keycloak upgrade.

It seems that this change needs a bit more discussion, therefore I'll move it to a separate PR.